src/net/libwsIO.cpp

```diff
@@ -0,0 +1,213 @@
+#include "net/libwsIO.h"
+#include <arpa/inet.h>
+#include <libws_log.h>
+#include <gcmpp.h>
+
+#include "waiter/libeventWaiter.h"
+
+using namespace std;
+
+LibwsIO::LibwsIO(::mega::Mutex *mutex, void *ctx) : WebsocketsIO(mutex, ctx)
+{
+    initialized = false;
+}
+
+LibwsIO::~LibwsIO()
+{
+    // ws_global_destroy() is not consistent with ws_global_init()
+    // ws_global_init() expect wscontext to be externally allocated, but
+    // ws_global_destroy() expect to delete it using free()
+}
+
+void LibwsIO::addevents(::mega::Waiter* waiter, int)
+{
+    if (!initialized)
+    {
+        ::mega::LibeventWaiter *libeventWaiter = dynamic_cast<::mega::LibeventWaiter *>(waiter);
+        ws_global_init(&wscontext, libeventWaiter ? libeventWaiter->eventloop : services_get_event_loop(), NULL,
+        [](struct bufferevent* bev, void* userp)
+        {
+            karere::marshallCall([bev, userp]()
+            {
+                ws_read_callback(bev, userp);
+            }, NULL);
+        },
+        [](struct bufferevent* bev, short events, void* userp)
+        {
+            karere::marshallCall([bev, events, userp]()
+            {
+                ws_event_callback(bev, events, userp);
+            }, NULL);
+        },
+        [](int fd, short events, void* userp)
+        {
+            karere::marshallCall([events, userp]()
+            {
+                ws_handle_marshall_timer_cb(0, events, userp);
+            }, NULL);
+        });
+        //ws_set_log_level(LIBWS_TRACE);
+        initialized = true;
+    }
+}
+
+WebsocketsClientImpl *LibwsIO::wsConnect(const char *ip, const char *host, int port, const char *path, bool ssl, WebsocketsClient *client)
+{
+    if (!initialized)
+    {
+        addevents(NULL, 0);
```

Comment:
This check is required because the Qt app doesn't use the intermediate layer and therefore it doesn't explicitly call `addevents()`. Hence, Qt app will always enter into the block, while apps based on MegaChatApi will never execute it.